### ISSUE TITLE
Throw a better exception on error

### ISSUE THREAD

**Issue by jbattermann** [mj1856/SimpleImpersonation#14]
Currently whenever impersonation cannot be performed / completed an [System.ApplicationException](https://github.com/mj1856/SimpleImpersonation/blob/master/SimpleImpersonation/Impersonation.cs#L84) is thrown with the [underlying error code](https://github.com/mj1856/SimpleImpersonation/blob/master/SimpleImpersonation/Impersonation.cs#L77) only available inside the generic string.. which makes it quite hard from a developer point of view to find out what exactly is wrong and handle it appropriately.

It would be neat to get the underlying cause when using the library, i.e. via throwing a [Win32Exception](https://msdn.microsoft.com/en-us/library/system.componentmodel.win32exception(v=vs.110).aspx) instead (I'll send over a PR shortly).

**Comment by mj1856** [comment]
While I agree that it would be good to have more direct access to the error reason, I think throwing a `Win32Exception` is the wrong way to go about it.  That still requires the end-user to have knowledge of the integer error codes.  Also, it would be a breaking change for those that are already implementing it and catching `ApplicationException`.

Perhaps a better plan would be to include an enumeration of errors, and a custom exception class that provides the enumeration.  If the exception class inherits `ApplicationException`, then it would be a non-breaking change.

```csharp
public class ImpersonationException : ApplicationException
{
   // ctor...

   public ImpersonationError Error { get; private set; }
}

public enum ImpersonationError
{
  // ...
}
```

Of course, we would need to research all of the possible error codes that could be returned, but I think it is appropriate to do that here, rather than asking the consumer of this library to figure them out.

**Comment by jbattermann** [comment]
Hi @mj1856 the neat thing about Win32Exception is/was that it does resolve the Windows Error codes automatically - i.e. in case of incorrect username or password (error code 1326), the Win32Exception does include the appropriate, human-readable and localized error .Message.

Mapping all the possibilities [1326, 1328, 1329, 1330, ...](https://msdn.microsoft.com/en-us/library/windows/desktop/ms681385(v=vs.85).aspx) to enums / custom error & localized messages is quite an effort.

So if you want to roll your own Exception I'd personally be happy with a mere int .Error, .ErrorId or .ErrorCode in it.

**Comment by mj1856** [comment]
Interesting.  I wasn't aware of that little tidbit.  Thanks!

I'm now debating on whether it's more important to not break compatibility, or to get the error message out correctly.  There are about 11,000 downloads of this package on Nuget, and we don't have a very well defined communications channel to announce breaking changes.  So I can see folks getting a nasty surprise if they're catching `ApplicationException` and it stops working after an upgrade.

Here's an idea that might solve both concerns.

```csharp
public class ImpersonationException : ApplicationException
{
    public ImpersonationException(Win32Exception win32Exception)
        : base(win32Exception.Message, win32Exception) { }

    public int ErrorCode => ((Win32Exception)InnerException).ErrorCode;
}
```

That allows the error code to easily be retrieved, uses the message from the OS, and the caller can get at the `Win32Exception` via the `InnerException` property.

It does mean that anyone parsing the error code out of the existing message string would still have a breaking change, but I think that's less critical than missing the exception entirely.

What do you think?

**Comment by jbattermann** [comment]
Looks absolutely fine (actually quite good) to me 👍 Regarding the remaining issue of the actual .Message content change & users potentially parsing the error code out etc: you could also bump the major version of the NuGet package to 2.x.

But either way - it does look good that way!

**Comment by mj1856** [comment]
Yep. I'll do that and mark it 2.0.0 just to be strictly SemVer compliant.

**Comment by jbattermann** [comment]
👍

**Comment by jbattermann** [comment]
Nice, thanks @mj1856!

**Comment by mj1856** [comment]
Released 2.0.0.  LMK if you encounter any difficulties.

And thanks!
